### ISSUE TITLE
Implement localization for UI labels

### ISSUE THREAD

**Issue by misfir3** [WebGoat/WebGoat#266]
Implement general js support (model) & implement labels in UI with localized labels

**Comment by zupzup** [comment]
Hey,

I'd like to try myself at this if I may. I did some analysis in order to come up with an attack plan for this issue and came up with some questions, which would probably have to be answered before starting:

**On plugins:**
One issue which I already encountered somewhat in https://github.com/WebGoat/WebGoat/issues/265 was that currently, plugins are not properly localized, as they only use the "plugin/i18n/WebGoatLabels.properties" file and not the other languages as seen [here](https://github.com/WebGoat/WebGoat/blob/develop/webgoat-container/src/main/java/org/owasp/webgoat/plugins/PluginsExtractor.java#L153-L154)
So that probably needs fixing.

Moreover, the question how plugins are to be localized presents itself, as the labels from the webgoat-container and all the plugins are merged together, a name-clash between plugins or the container would lead to labels being overwritten.

I think namespacing would be a viable route here, either by convention (e.g.: people need to use the prefix "xy" for labels in plugin "xy") or automatically prefixing the labels of a plugin - but in the UI, people would need to use the correctly prefixed label anyways.

Another possibility would be to create a specific labels-map for each plugin, then the UI could use keys without prefixes, as long as it uses the correct labels-map and fall back to the webgoat-container labels.

**On the UI side:**
I guess it should suffice to create a LabelsModel which is used by Views that use labels. That LabelsModel would just use the LabelService on the backend (and probably cache the data somehow, because they don't change regularly).
One could create something like a LabelView to extend, but I think that would probably make things more complex that they need to be.

Am I thinking too simple here, or would this be an acceptable approach?

**On labels in general:**
I'm not quite sure what qualifies as a label to be localized. Is it just every piece of text in the whole application?

E.g.: Menu, User Controls, Everything within lessons...

And what about the login-page - it is kind of external to the whole Backbone app, so that would probably have to be localized another way, right?

That's it :) I think if these things are cleared up, the actual implementation shouldn't be a lot of trouble.

**Comment by misfir3** [comment]
@zupzup Thanks for looking into this and your thoughts (and all your contributions so far)

The intent with this (as I try to remember back)  as 'labels' was really for UI controls.  Anything that will be present in the UI (may have conditional hide/show) and have the same label whenever it is shown (e.g. hints).

I'm going to create a separate issue for lesson/content localization (in the *.html files).

Also, we really need to determine which (initial) buttons/controls that will need labels. E.g. I know we need to revisit hints and how we plan to implement them with the new platform. So, while I thought I was going to further this myself, I realize we need to do a few other things first.

Will email the list or send something out on Slack with a suggested prioritization.

**Comment by nbaars** [comment]
Closing the issue as the UI is now using this endpoint for translation
